File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/TargetKindResolver.java
package org.jetbrains.bsp.bazel.server.sync;

import io.vavr.collection.HashSet;
import io.vavr.collection.Set;
import java.util.Map;
import org.jetbrains.bsp.bazel.info.BspTargetInfo.TargetInfo;
import org.jetbrains.bsp.bazel.server.sync.model.Tag;

public class TargetKindResolver {
  private static final Set<Tag> LIBRARY = HashSet.of(Tag.LIBRARY);
  private static final Map<String, Set<Tag>> ruleSuffixToTargetType =
      Map.of(
          "library", LIBRARY,
          "binary", HashSet.of(Tag.APPLICATION),
          "test", HashSet.of(Tag.TEST));

  private static final Set<Tag> NO_IDE = HashSet.of(Tag.NO_IDE);

  public Set<Tag> resolveTags(TargetInfo targetInfo) {
    if (targetInfo.getKind().equals("resources_union")) {
      return LIBRARY;
    }

    var tag =
        ruleSuffixToTargetType.entrySet().stream()
            .filter(entry -> targetInfo.getKind().endsWith("_" + entry.getKey()))
            .map(Map.Entry::getValue)
            .findFirst()
            .orElse(NO_IDE);
    if (targetInfo.getTagsList().contains("no-ide")) {
      return tag.add(Tag.NO_IDE);
    }
    return tag;
  }
}

File: server/src/main/java/org/jetbrains/bsp/bazel/server/sync/TargetKindResolver.kt
package org.jetbrains.bsp.bazel.server.sync

import org.jetbrains.bsp.bazel.info.BspTargetInfo
import org.jetbrains.bsp.bazel.server.sync.model.Tag

class TargetKindResolver {
    fun resolveTags(targetInfo: BspTargetInfo.TargetInfo): Set<Tag> {
        if (targetInfo.kind == "resources_union" || targetInfo.kind == "java_import") {
            return LIBRARY
        }
        val tag = ruleSuffixToTargetType.filterKeys {
            targetInfo.kind.endsWith("_$it")
        }.values.firstOrNull() ?: NO_IDE
        return if (targetInfo.tagsList.contains("no-ide")) {
            tag + Tag.NO_IDE
        } else if (targetInfo.tagsList.contains("manual")) {
            tag + Tag.MANUAL
        } else tag
    }

    companion object {
        private val LIBRARY: Set<Tag> = hashSetOf(Tag.LIBRARY)
        private val ruleSuffixToTargetType = java.util.Map.of(
            "library", LIBRARY,
            "binary", hashSetOf(Tag.APPLICATION),
            "test", hashSetOf(Tag.TEST)
        )
        private val NO_IDE: Set<Tag> = hashSetOf(Tag.NO_IDE)
    }
}
